lib/messageSender.ts

```diff
@@ -0,0 +1,523 @@
+          this._sender!.removeListener(SenderEvents.accepted, onAccepted);
+          this._sender!.removeListener(SenderEvents.released, onReleased);
+          this._sender!.removeListener(SenderEvents.modified, onModified);
+        };
+
+        onAccepted = (context: EventContext) => {
+          // Since we will be adding listener for accepted and rejected event every time
+          // we send a message, we need to remove listener for both the events.
+          // This will ensure duplicate listeners are not added for the same event.
+          removeListeners();
+          log.sender("[%s] Sender '%s', got event accepted.",
+            this._context.namespace.connectionId, this.name);
+          resolve(context.delivery);
+        };
+        onRejected = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event rejected.",
+            this._context.namespace.connectionId, this.name);
+          const err = translate(context!.delivery!.remote_state!.error);
+          log.error(err);
+          reject(err);
+        };
+        onReleased = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event released.",
+            this._context.namespace.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(`[${this._context.namespace.connectionId}]Sender '${this.name}', ` +
+              `received a release disposition.Hence we are rejecting the promise.`);
+          }
+          log.error(err);
+          reject(err);
+        };
+        onModified = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event modified.",
+            this._context.namespace.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(`[${this._context.namespace.connectionId}]Sender "${this.name}", ` +
+              `received a modified disposition.Hence we are rejecting the promise.`);
+          }
+          log.error(err);
+          reject(err);
+        };
+
+        const actionAfterTimeout = () => {
+          removeListeners();
+          const desc: string = `[${this._context.namespace.connectionId}] Sender "${this.name}" ` +
+            `with address "${this.address}", was not able to send the message right now, due ` +
+            `to operation timeout.`;
+          log.error(desc);
+          const e: AmqpError = {
+            condition: ErrorNameConditionMapper.ServiceUnavailableError,
+            description: desc
+          };
+          return reject(translate(e));
+        };
+
+        this._sender!.on(SenderEvents.accepted, onAccepted);
```

Comment:
these will still run up against the event emitter warnings for listener count right?

Reply:
Yup. I have set the max limit to 1000 for now.
```
this._sender.setMaxListeners(1000);
```